# Introduction

This notebook contains two parts. **Part 1, Testing Perceptrons**, provides you an opportunity to demonstrate your ability to apply course concepts by implementing a test function for a perceptron. **Part 2, Mine Detection**, provides you an opportunity to practice using widely-used ML libraries and an ML workflow to solve a classification problem.

You do not need to complete Part 1 in order to complete Part 2. If you get stuck on Part 1, and choose to work on Part 2, be sure that all of your code for Part 1 runs without error. You can comment out your code in Part 1 if necessary.

# Part 1: Testing Perceptrons

Given a simple Perceptron classifier, and a test set cancer diagnoses, demonstrate your ability to implement a perceptron's `test` function, such that it returns the accuracy of its predictions for the class labels of samples in a training set.

## The Perceptron Implementation

Let's first introduce the classifier, which you should find familiar, and you do not need to modify. Notice that the `test` method is stubbed.

In [39]:
import random

class Perceptron:

    def __init__(self, alpha = 0.1, max_epochs = 1000):
        self.alpha = alpha
        self.max_epochs = max_epochs

    def train(self, training_set):
        self.weights = self._initial_weights(len(training_set[0]))
        for i in range(self.max_epochs):
            correct_classifications = 0
            for record in training_set:
                y = record[0]
                y_hat = self.predict(record[1:])
                if y == y_hat: correct_classifications += 1
                self._update_weights(y - y_hat, [1] + record[1:])
            if correct_classifications / len(training_set) == 1.0:
                print(f"Epoch {i} Accuracy: {correct_classifications / len(training_set)}")
                break

    def predict(self, features):
        return self._sign_of(self._dot_product(self.weights, [1] + features))

    def test(self, test_set):
        pass

    def _update_weights(self, error, features):
        self.weights[0] += self.alpha * error
        for i in range(1, len(self.weights)):
            self.weights[i] += self.alpha * error * features[i]

    def _dot_product(self, w, x):
        return sum([w * x for w, x in zip(w, x)])

    def _sign_of(self, value):
        return 1 if value >= 0 else -1

    def _initial_weights(self, length):
        return [random.uniform(0, 1) for _ in range(length)]

## The Data Set

There is no need for you to manually load the data set. We have provided a subset of the cancer diagnoses data set here, already split into a training set, `diagnoses_training`, and a test set, `diagnoses_test`. Each data set is a simple, two-dimensional Python list, where each sub-list represents the attributes for one diagnosis.

We have preprocessed the data, such that the **first dimension is the class label**: a `1` indicating malignant, and `-1` indicating benign.


In [40]:
diagnoses_training = [
    [1, 17.99, 10.38, 122.8, 1001, 0.1184, 0.2776, 0.3001, 0.1471, 0.2419, 0.07871, 1.095, 0.9053, 8.589, 153.4, 0.006399, 0.04904, 0.05373, 0.01587, 0.03003, 0.006193, 25.38, 17.33, 184.6, 2019, 0.1622, 0.6656, 0.7119, 0.2654, 0.4601, 0.1189],
    [1, 20.57, 17.77, 132.9, 1326, 0.08474, 0.07864, 0.0869, 0.07017, 0.1812, 0.05667, 0.5435, 0.7339, 3.398, 74.08, 0.005225, 0.01308, 0.0186, 0.0134, 0.01389, 0.003532, 24.99, 23.41, 158.8, 1956, 0.1238, 0.1866, 0.2416, 0.186, 0.275, 0.08902],
    [1, 19.69, 21.25, 130, 1203, 0.1096, 0.1599, 0.1974, 0.1279, 0.2069, 0.05999, 0.7456, 0.7869, 4.585, 94.03, 0.00615, 0.04006, 0.03832, 0.02058, 0.0225, 0.004571, 23.57, 25.53, 152.5, 1709, 0.1444, 0.4245, 0.4504, 0.243, 0.3613, 0.08758],
    [1, 11.42, 20.38, 77.58, 386.1, 0.1425, 0.2839, 0.2414, 0.1052, 0.2597, 0.09744, 0.4956, 1.156, 3.445, 27.23, 0.00911, 0.07458, 0.05661, 0.01867, 0.05963, 0.009208, 14.91, 26.5, 98.87, 567.7, 0.2098, 0.8663, 0.6869, 0.2575, 0.6638, 0.173],
    [1, 20.29, 14.34, 135.1, 1297, 0.1003, 0.1328, 0.198, 0.1043, 0.1809, 0.05883, 0.7572, 0.7813, 5.438, 94.44, 0.01149, 0.02461, 0.05688, 0.01885, 0.01756, 0.005115, 22.54, 16.67, 152.2, 1575, 0.1374, 0.205, 0.4, 0.1625, 0.2364, 0.07678],
    [1, 12.45, 15.7, 82.57, 477.1, 0.1278, 0.17, 0.1578, 0.08089, 0.2087, 0.07613, 0.3345, 0.8902, 2.217, 27.19, 0.00751, 0.03345, 0.03672, 0.01137, 0.02165, 0.005082, 15.47, 23.75, 103.4, 741.6, 0.1791, 0.5249, 0.5355, 0.1741, 0.3985, 0.1244],
    [1, 18.25, 19.98, 119.6, 1040, 0.09463, 0.109, 0.1127, 0.074, 0.1794, 0.05742, 0.4467, 0.7732, 3.18, 53.91, 0.004314, 0.01382, 0.02254, 0.01039, 0.01369, 0.002179, 22.88, 27.66, 153.2, 1606, 0.1442, 0.2576, 0.3784, 0.1932, 0.3063, 0.08368],
    [1, 13.71, 20.83, 90.2, 577.9, 0.1189, 0.1645, 0.09366, 0.05985, 0.2196, 0.07451, 0.5835, 1.377, 3.856, 50.96, 0.008805, 0.03029, 0.02488, 0.01448, 0.01486, 0.005412, 17.06, 28.14, 110.6, 897, 0.1654, 0.3682, 0.2678, 0.1556, 0.3196, 0.1151],
    [1, 13, 21.82, 87.5, 519.8, 0.1273, 0.1932, 0.1859, 0.09353, 0.235, 0.07389, 0.3063, 1.002, 2.406, 24.32, 0.005731, 0.03502, 0.03553, 0.01226, 0.02143, 0.003749, 15.49, 30.73, 106.2, 739.3, 0.1703, 0.5401, 0.539, 0.206, 0.4378, 0.1072],
    [1, 12.46, 24.04, 83.97, 475.9, 0.1186, 0.2396, 0.2273, 0.08543, 0.203, 0.08243, 0.2976, 1.599, 2.039, 23.94, 0.007149, 0.07217, 0.07743, 0.01432, 0.01789, 0.01008, 15.09, 40.68, 97.65, 711.4, 0.1853, 1.058, 1.105, 0.221, 0.4366, 0.2075],
    [1, 16.02, 23.24, 102.7, 797.8, 0.08206, 0.06669, 0.03299, 0.03323, 0.1528, 0.05697, 0.3795, 1.187, 2.466, 40.51, 0.004029, 0.009269, 0.01101, 0.007591, 0.0146, 0.003042, 19.19, 33.88, 123.8, 1150, 0.1181, 0.1551, 0.1459, 0.09975, 0.2948, 0.08452],
    [1, 15.78, 17.89, 103.6, 781, 0.0971, 0.1292, 0.09954, 0.06606, 0.1842, 0.06082, 0.5058, 0.9849, 3.564, 54.16, 0.005771, 0.04061, 0.02791, 0.01282, 0.02008, 0.004144, 20.42, 27.28, 136.5, 1299, 0.1396, 0.5609, 0.3965, 0.181, 0.3792, 0.1048],
    [1, 19.17, 24.8, 132.4, 1123, 0.0974, 0.2458, 0.2065, 0.1118, 0.2397, 0.078, 0.9555, 3.568, 11.07, 116.2, 0.003139, 0.08297, 0.0889, 0.0409, 0.04484, 0.01284, 20.96, 29.94, 151.7, 1332, 0.1037, 0.3903, 0.3639, 0.1767, 0.3176, 0.1023],
    [1, 15.85, 23.95, 103.7, 782.7, 0.08401, 0.1002, 0.09938, 0.05364, 0.1847, 0.05338, 0.4033, 1.078, 2.903, 36.58, 0.009769, 0.03126, 0.05051, 0.01992, 0.02981, 0.003002, 16.84, 27.66, 112, 876.5, 0.1131, 0.1924, 0.2322, 0.1119, 0.2809, 0.06287],
    [1, 13.73, 22.61, 93.6, 578.3, 0.1131, 0.2293, 0.2128, 0.08025, 0.2069, 0.07682, 0.2121, 1.169, 2.061, 19.21, 0.006429, 0.05936, 0.05501, 0.01628, 0.01961, 0.008093, 15.03, 32.01, 108.8, 697.7, 0.1651, 0.7725, 0.6943, 0.2208, 0.3596, 0.1431],
    [1, 14.54, 27.54, 96.73, 658.8, 0.1139, 0.1595, 0.1639, 0.07364, 0.2303, 0.07077, 0.37, 1.033, 2.879, 32.55, 0.005607, 0.0424, 0.04741, 0.0109, 0.01857, 0.005466, 17.46, 37.13, 124.1, 943.2, 0.1678, 0.6577, 0.7026, 0.1712, 0.4218, 0.1341],
    [1, 14.68, 20.13, 94.74, 684.5, 0.09867, 0.072, 0.07395, 0.05259, 0.1586, 0.05922, 0.4727, 1.24, 3.195, 45.4, 0.005718, 0.01162, 0.01998, 0.01109, 0.0141, 0.002085, 19.07, 30.88, 123.4, 1138, 0.1464, 0.1871, 0.2914, 0.1609, 0.3029, 0.08216],
    [1, 16.13, 20.68, 108.1, 798.8, 0.117, 0.2022, 0.1722, 0.1028, 0.2164, 0.07356, 0.5692, 1.073, 3.854, 54.18, 0.007026, 0.02501, 0.03188, 0.01297, 0.01689, 0.004142, 20.96, 31.48, 136.8, 1315, 0.1789, 0.4233, 0.4784, 0.2073, 0.3706, 0.1142],
    [1, 19.81, 22.15, 130, 1260, 0.09831, 0.1027, 0.1479, 0.09498, 0.1582, 0.05395, 0.7582, 1.017, 5.865, 112.4, 0.006494, 0.01893, 0.03391, 0.01521, 0.01356, 0.001997, 27.32, 30.88, 186.8, 2398, 0.1512, 0.315, 0.5372, 0.2388, 0.2768, 0.07615],
    [-1, 13.54, 14.36, 87.46, 566.3, 0.09779, 0.08129, 0.06664, 0.04781, 0.1885, 0.05766, 0.2699, 0.7886, 2.058, 23.56, 0.008462, 0.0146, 0.02387, 0.01315, 0.0198, 0.0023, 15.11, 19.26, 99.7, 711.2, 0.144, 0.1773, 0.239, 0.1288, 0.2977, 0.07259],
    [-1, 13.08, 15.71, 85.63, 520, 0.1075, 0.127, 0.04568, 0.0311, 0.1967, 0.06811, 0.1852, 0.7477, 1.383, 14.67, 0.004097, 0.01898, 0.01698, 0.00649, 0.01678, 0.002425, 14.5, 20.49, 96.09, 630.5, 0.1312, 0.2776, 0.189, 0.07283, 0.3184, 0.08183],
    [-1, 9.504, 12.44, 60.34, 273.9, 0.1024, 0.06492, 0.02956, 0.02076, 0.1815, 0.06905, 0.2773, 0.9768, 1.909, 15.7, 0.009606, 0.01432, 0.01985, 0.01421, 0.02027, 0.002968, 10.23, 15.66, 65.13, 314.9, 0.1324, 0.1148, 0.08867, 0.06227, 0.245, 0.07773],
    [1, 15.34, 14.26, 102.5, 704.4, 0.1073, 0.2135, 0.2077, 0.09756, 0.2521, 0.07032, 0.4388, 0.7096, 3.384, 44.91, 0.006789, 0.05328, 0.06446, 0.02252, 0.03672, 0.004394, 18.07, 19.08, 125.1, 980.9, 0.139, 0.5954, 0.6305, 0.2393, 0.4667, 0.09946],
    [1, 21.16, 23.04, 137.2, 1404, 0.09428, 0.1022, 0.1097, 0.08632, 0.1769, 0.05278, 0.6917, 1.127, 4.303, 93.99, 0.004728, 0.01259, 0.01715, 0.01038, 0.01083, 0.001987, 29.17, 35.59, 188, 2615, 0.1401, 0.26, 0.3155, 0.2009, 0.2822, 0.07526],
    [1, 16.65, 21.38, 110, 904.6, 0.1121, 0.1457, 0.1525, 0.0917, 0.1995, 0.0633, 0.8068, 0.9017, 5.455, 102.6, 0.006048, 0.01882, 0.02741, 0.0113, 0.01468, 0.002801, 26.46, 31.56, 177, 2215, 0.1805, 0.3578, 0.4695, 0.2095, 0.3613, 0.09564],
    [1, 17.14, 16.4, 116, 912.7, 0.1186, 0.2276, 0.2229, 0.1401, 0.304, 0.07413, 1.046, 0.976, 7.276, 111.4, 0.008029, 0.03799, 0.03732, 0.02397, 0.02308, 0.007444, 22.25, 21.4, 152.4, 1461, 0.1545, 0.3949, 0.3853, 0.255, 0.4066, 0.1059],
    [1, 14.58, 21.53, 97.41, 644.8, 0.1054, 0.1868, 0.1425, 0.08783, 0.2252, 0.06924, 0.2545, 0.9832, 2.11, 21.05, 0.004452, 0.03055, 0.02681, 0.01352, 0.01454, 0.003711, 17.62, 33.21, 122.4, 896.9, 0.1525, 0.6643, 0.5539, 0.2701, 0.4264, 0.1275],
    [1, 18.61, 20.25, 122.1, 1094, 0.0944, 0.1066, 0.149, 0.07731, 0.1697, 0.05699, 0.8529, 1.849, 5.632, 93.54, 0.01075, 0.02722, 0.05081, 0.01911, 0.02293, 0.004217, 21.31, 27.26, 139.9, 1403, 0.1338, 0.2117, 0.3446, 0.149, 0.2341, 0.07421],
    [1, 15.3, 25.27, 102.4, 732.4, 0.1082, 0.1697, 0.1683, 0.08751, 0.1926, 0.0654, 0.439, 1.012, 3.498, 43.5, 0.005233, 0.03057, 0.03576, 0.01083, 0.01768, 0.002967, 20.27, 36.71, 149.3, 1269, 0.1641, 0.611, 0.6335, 0.2024, 0.4027, 0.09876],
    [1, 17.57, 15.05, 115, 955.1, 0.09847, 0.1157, 0.09875, 0.07953, 0.1739, 0.06149, 0.6003, 0.8225, 4.655, 61.1, 0.005627, 0.03033, 0.03407, 0.01354, 0.01925, 0.003742, 20.01, 19.52, 134.9, 1227, 0.1255, 0.2812, 0.2489, 0.1456, 0.2756, 0.07919],
    [1, 18.63, 25.11, 124.8, 1088, 0.1064, 0.1887, 0.2319, 0.1244, 0.2183, 0.06197, 0.8307, 1.466, 5.574, 105, 0.006248, 0.03374, 0.05196, 0.01158, 0.02007, 0.00456, 23.15, 34.01, 160.5, 1670, 0.1491, 0.4257, 0.6133, 0.1848, 0.3444, 0.09782],
    [1, 11.84, 18.7, 77.93, 440.6, 0.1109, 0.1516, 0.1218, 0.05182, 0.2301, 0.07799, 0.4825, 1.03, 3.475, 41, 0.005551, 0.03414, 0.04205, 0.01044, 0.02273, 0.005667, 16.82, 28.12, 119.4, 888.7, 0.1637, 0.5775, 0.6956, 0.1546, 0.4761, 0.1402],
    [1, 17.02, 23.98, 112.8, 899.3, 0.1197, 0.1496, 0.2417, 0.1203, 0.2248, 0.06382, 0.6009, 1.398, 3.999, 67.78, 0.008268, 0.03082, 0.05042, 0.01112, 0.02102, 0.003854, 20.88, 32.09, 136.1, 1344, 0.1634, 0.3559, 0.5588, 0.1847, 0.353, 0.08482],
    [1, 19.27, 26.47, 127.9, 1162, 0.09401, 0.1719, 0.1657, 0.07593, 0.1853, 0.06261, 0.5558, 0.6062, 3.528, 68.17, 0.005015, 0.03318, 0.03497, 0.009643, 0.01543, 0.003896, 24.15, 30.9, 161.4, 1813, 0.1509, 0.659, 0.6091, 0.1785, 0.3672, 0.1123],
    [1, 16.13, 17.88, 107, 807.2, 0.104, 0.1559, 0.1354, 0.07752, 0.1998, 0.06515, 0.334, 0.6857, 2.183, 35.03, 0.004185, 0.02868, 0.02664, 0.009067, 0.01703, 0.003817, 20.21, 27.26, 132.7, 1261, 0.1446, 0.5804, 0.5274, 0.1864, 0.427, 0.1233],
    [1, 16.74, 21.59, 110.1, 869.5, 0.0961, 0.1336, 0.1348, 0.06018, 0.1896, 0.05656, 0.4615, 0.9197, 3.008, 45.19, 0.005776, 0.02499, 0.03695, 0.01195, 0.02789, 0.002665, 20.01, 29.02, 133.5, 1229, 0.1563, 0.3835, 0.5409, 0.1813, 0.4863, 0.08633],
    [1, 14.25, 21.72, 93.63, 633, 0.09823, 0.1098, 0.1319, 0.05598, 0.1885, 0.06125, 0.286, 1.019, 2.657, 24.91, 0.005878, 0.02995, 0.04815, 0.01161, 0.02028, 0.004022, 15.89, 30.36, 116.2, 799.6, 0.1446, 0.4238, 0.5186, 0.1447, 0.3591, 0.1014],
    [-1, 13.03, 18.42, 82.61, 523.8, 0.08983, 0.03766, 0.02562, 0.02923, 0.1467, 0.05863, 0.1839, 2.342, 1.17, 14.16, 0.004352, 0.004899, 0.01343, 0.01164, 0.02671, 0.001777, 13.3, 22.81, 84.46, 545.9, 0.09701, 0.04619, 0.04833, 0.05013, 0.1987, 0.06169],
    [1, 14.99, 25.2, 95.54, 698.8, 0.09387, 0.05131, 0.02398, 0.02899, 0.1565, 0.05504, 1.214, 2.188, 8.077, 106, 0.006883, 0.01094, 0.01818, 0.01917, 0.007882, 0.001754, 14.99, 25.2, 95.54, 698.8, 0.09387, 0.05131, 0.02398, 0.02899, 0.1565, 0.05504],
    [1, 13.48, 20.82, 88.4, 559.2, 0.1016, 0.1255, 0.1063, 0.05439, 0.172, 0.06419, 0.213, 0.5914, 1.545, 18.52, 0.005367, 0.02239, 0.03049, 0.01262, 0.01377, 0.003187, 15.53, 26.02, 107.3, 740.4, 0.161, 0.4225, 0.503, 0.2258, 0.2807, 0.1071],
    [1, 13.44, 21.58, 86.18, 563, 0.08162, 0.06031, 0.0311, 0.02031, 0.1784, 0.05587, 0.2385, 0.8265, 1.572, 20.53, 0.00328, 0.01102, 0.0139, 0.006881, 0.0138, 0.001286, 15.93, 30.25, 102.5, 787.9, 0.1094, 0.2043, 0.2085, 0.1112, 0.2994, 0.07146],
    [1, 10.95, 21.35, 71.9, 371.1, 0.1227, 0.1218, 0.1044, 0.05669, 0.1895, 0.0687, 0.2366, 1.428, 1.822, 16.97, 0.008064, 0.01764, 0.02595, 0.01037, 0.01357, 0.00304, 12.84, 35.34, 87.22, 514, 0.1909, 0.2698, 0.4023, 0.1424, 0.2964, 0.09606],
    [1, 19.07, 24.81, 128.3, 1104, 0.09081, 0.219, 0.2107, 0.09961, 0.231, 0.06343, 0.9811, 1.666, 8.83, 104.9, 0.006548, 0.1006, 0.09723, 0.02638, 0.05333, 0.007646, 24.09, 33.17, 177.4, 1651, 0.1247, 0.7444, 0.7242, 0.2493, 0.467, 0.1038],
    [1, 13.28, 20.28, 87.32, 545.2, 0.1041, 0.1436, 0.09847, 0.06158, 0.1974, 0.06782, 0.3704, 0.8249, 2.427, 31.33, 0.005072, 0.02147, 0.02185, 0.00956, 0.01719, 0.003317, 17.38, 28, 113.1, 907.2, 0.153, 0.3724, 0.3664, 0.1492, 0.3739, 0.1027],
    [1, 13.17, 21.81, 85.42, 531.5, 0.09714, 0.1047, 0.08259, 0.05252, 0.1746, 0.06177, 0.1938, 0.6123, 1.334, 14.49, 0.00335, 0.01384, 0.01452, 0.006853, 0.01113, 0.00172, 16.23, 29.89, 105.5, 740.7, 0.1503, 0.3904, 0.3728, 0.1607, 0.3693, 0.09618],
    [1, 18.65, 17.6, 123.7, 1076, 0.1099, 0.1686, 0.1974, 0.1009, 0.1907, 0.06049, 0.6289, 0.6633, 4.293, 71.56, 0.006294, 0.03994, 0.05554, 0.01695, 0.02428, 0.003535, 22.82, 21.32, 150.6, 1567, 0.1679, 0.509, 0.7345, 0.2378, 0.3799, 0.09185],
    [-1, 8.196, 16.84, 51.71, 201.9, 0.086, 0.05943, 0.01588, 0.005917, 0.1769, 0.06503, 0.1563, 0.9567, 1.094, 8.205, 0.008968, 0.01646, 0.01588, 0.005917, 0.02574, 0.002582, 8.964, 21.96, 57.26, 242.2, 0.1297, 0.1357, 0.0688, 0.02564, 0.3105, 0.07409]
]

diagnoses_test = [
    [-1, 8.95, 15.76, 58.74, 245.2, 0.09462, 0.1243, 0.09263, 0.02308, 0.1305, 0.07163, 0.3132, 0.9789, 3.28, 16.94, 0.01835, 0.0676, 0.09263, 0.02308, 0.02384, 0.005601, 9.414, 17.07, 63.34, 270, 0.1179, 0.1879, 0.1544, 0.03846, 0.1652, 0.07722],
    [1, 15.22, 30.62, 103.4, 716.9, 0.1048, 0.2087, 0.255, 0.09429, 0.2128, 0.07152, 0.2602, 1.205, 2.362, 22.65, 0.004625, 0.04844, 0.07359, 0.01608, 0.02137, 0.006142, 17.52, 42.79, 128.7, 915, 0.1417, 0.7917, 1.17, 0.2356, 0.4089, 0.1409],
    [-1, 11.34, 21.26, 72.48, 396.5, 0.08759, 0.06575, 0.05133, 0.01899, 0.1487, 0.06529, 0.2344, 0.9861, 1.597, 16.41, 0.009113, 0.01557, 0.02443, 0.006435, 0.01568, 0.002477, 13.01, 29.15, 83.99, 518.1, 0.1699, 0.2196, 0.312, 0.08278, 0.2829, 0.08832],
    [1, 20.92, 25.09, 143, 1347, 0.1099, 0.2236, 0.3174, 0.1474, 0.2149, 0.06879, 0.9622, 1.026, 8.758, 118.8, 0.006399, 0.0431, 0.07845, 0.02624, 0.02057, 0.006213, 24.29, 29.41, 179.1, 1819, 0.1407, 0.4186, 0.6599, 0.2542, 0.2929, 0.09873],
    [-1, 12.36, 18.54, 79.01, 466.7, 0.08477, 0.06815, 0.02643, 0.01921, 0.1602, 0.06066, 0.1199, 0.8944, 0.8484, 9.227, 0.003457, 0.01047, 0.01167, 0.005558, 0.01251, 0.001356, 13.29, 27.49, 85.56, 544.1, 0.1184, 0.1963, 0.1937, 0.08442, 0.2983, 0.07185],
    [1, 21.56, 22.39, 142, 1479, 0.111, 0.1159, 0.2439, 0.1389, 0.1726, 0.05623, 1.176, 1.256, 7.673, 158.7, 0.0103, 0.02891, 0.05198, 0.02454, 0.01114, 0.004239, 25.45, 26.4, 166.1, 2027, 0.141, 0.2113, 0.4107, 0.2216, 0.206, 0.07115],
    [-1, 9.777, 16.99, 62.5, 290.2, 0.1037, 0.08404, 0.04334, 0.01778, 0.1584, 0.07065, 0.403, 1.424, 2.747, 22.87, 0.01385, 0.02932, 0.02722, 0.01023, 0.03281, 0.004638, 11.05, 21.47, 71.68, 367, 0.1467, 0.1765, 0.13, 0.05334, 0.2533, 0.08468],
    [1, 20.13, 28.25, 131.2, 1261, 0.0978, 0.1034, 0.144, 0.09791, 0.1752, 0.05533, 0.7655, 2.463, 5.203, 99.04, 0.005769, 0.02423, 0.0395, 0.01678, 0.01898, 0.002498, 23.69, 38.25, 155, 1731, 0.1166, 0.1922, 0.3215, 0.1628, 0.2572, 0.06637],
    [-1, 12.63, 20.76, 82.15, 480.4, 0.09933, 0.1209, 0.1065, 0.06021, 0.1735, 0.0707, 0.3424, 1.803, 2.711, 20.48, 0.01291, 0.04042, 0.05101, 0.02295, 0.02144, 0.005891, 13.33, 25.47, 89, 527.4, 0.1287, 0.225, 0.2216, 0.1105, 0.2226, 0.08486],
    [1, 16.6, 28.08, 108.3, 858.1, 0.08455, 0.1023, 0.09251, 0.05302, 0.159, 0.05648, 0.4564, 1.075, 3.425, 48.55, 0.005903, 0.03731, 0.0473, 0.01557, 0.01318, 0.003892, 18.98, 34.12, 126.7, 1124, 0.1139, 0.3094, 0.3403, 0.1418, 0.2218, 0.0782],
    [-1, 14.26, 19.65, 97.83, 629.9, 0.07837, 0.2233, 0.3003, 0.07798, 0.1704, 0.07769, 0.3628, 1.49, 3.399, 29.25, 0.005298, 0.07446, 0.1435, 0.02292, 0.02566, 0.01298, 15.3, 23.73, 107, 709, 0.08949, 0.4193, 0, 0.1503, 0.07247, 0.2438, 0.08541],
    [1, 20.6, 29.33, 140.1, 1265, 0.1178, 0.277, 0.3514, 0.152, 0.2397, 0.07016, 0.726, 1.595, 5.772, 86.22, 0.006522, 0.06158, 0.07117, 0.01664, 0.02324, 0.006185, 25.74, 39.42, 184.6, 1821, 0.165, 0.8681, 0.9387, 0.265, 0.4087, 0.124]
]


## What to Do
 
Demonstrate your understanding and ability to have synthesized course concepts by implementing a `test` function for the perceptron. Your goal is to implement the `test` function, stubbed for you below, in the subclass `TestablePerceptron`. The `test` function should return the accuracy rate of the perceptron's prediction, given `diagnoses_test`. In the end, your work should reflect the principles seen thus far in the course.

Please be sure to demonstrate:

1. Your implementation, as code, in the subclass below.
2. Using your test function, which is already done for you, following the class definition.

## Tips

- Be sure that you have spent time with the Exploration materials in this course.
- Ask questions on the course forum if you get stuck (describe what you are trying to do, and errors that you encounter)
- Keep it simple. This is quite straightforward.
- Be sure to run the code cell containing your TestablePerceptron class and the code cell that invokes the `test` method, or use the *>> Run All* button.

# Begin Your Implementation Here
In the code cell below, implement the test function by replacing pass with your code. It should accept a training set as input, and return a number representing the accuracy of the perceptron based on predictions made with the test set.

from exploration of material
> Our Perceptron also does not include an implementation of its test method, and we will not illustrate that here. Such a test method should:

> Iterate over every sample in the test set

> Make a prediction

> Keep track of the accuracy

> Report the result

> Notice that testing a perceptron is not training a perceptron, and that testing involves no training epochs, and no adjustment to a perceptron's weights.

In [41]:
class TestablePerceptron(Perceptron):
    
    def test(self, test_set):
        
        #creates varriable to store accurate predictions
        acc_pred = 0
    
        #Iterate over every sample in the test set and extracts dep. and ind. vars.
        for data_line in test_set:
            y = data_line[0]
            features = data_line[1:0]
        
            #Make a prediction and stores the result if correct
            y_hat = self.predict(features)
            if y == y_hat:
                acc_pred += 1
    
        #Report and return the results
        accuracy = acc_pred / len(test_set)
        return accuracy

    

Your TestablePerceptron should result in the following code, which trains and tests a TestablePerceptron, running without error. You do not need to modify the code below, but once your `test` method above is complete, you should see the accuracy of the test printed, instead of `None`.

In [42]:
perceptron = TestablePerceptron(alpha = 0.1, max_epochs = 10000)
perceptron.train(diagnoses_training)
print(perceptron.test(diagnoses_test))

Epoch 4982 Accuracy: 1.0
0.5


## Conclusion of Part 1

Write a few sentences here that describes what your `test` function does in "plain English." Try expressing this accurately and authentically *without* giving us a line-by-line summary of what your function does (we can read the code for that). Tell us what the test error is. Then, describe how many learning epochs it took for the perceptron to achive 100% accuracy during *training*, and explain why that number changes each time you run the code cell above.

**The test function asses the accuracy of the method’s ability to correctly determine if a cell is malignant or benign given a specific set of data. It uses numerical descriptions of each cell as the data input in order to make a prediction of malignant vs benign as the output. What the test function does is evaluate this prediction to see what percentage of correct predictions were made out of the entire set of data used to test the method/model.**

**The initial weights of the perceptron and generated randomly so there will be slight variations in the number of epochs required for convergence to 100% accuracy. In this case 5694 epochs were required.**



# Part 2: Mine Detection

In this, the second, part of this notebook, you will build a classifier that can predict whether or not a sonar signature is from a mine or a rock. We'll use a version of the [sonar data set](https://www.openml.org/search?type=data&sort=runs&id=40&status=active) by Gorman and Sejnowski. Take a moment now to [familiarize yourself with the subject matter of this data set](https://datahub.io/machine-learning/sonar%23resource-sonar), and look at the details of the version of this data set, [Mines vs Rocks, hosted on Kaggle](https://www.kaggle.com/datasets/mattcarter865/mines-vs-rocks).

Unlike previous notebooks, where we provide code for each step of the ML process, this notebook expects each student to implement the ML workflow steps. We will get you started by providing the first step, loading the data, and providing some landmarks below. Your process should demonstrate:

1. Loading the data
2. Exploring the data
3. Preprocessing the data
4. Preparing the training and test sets
5. Creating and configuring a sklearn.linear_model.Perceptron
6. Training the perceptron
7. Testing the perceptron
8. Demonstrating making predictions
9. Evaluate (and Improve) the results

Can you train a classifier that can predict whether a sonar signature is from a mine or a rock? "Three trained human subjects were each tested on 100 signals, chosen at random from the set of 208 returns used to create this data set. Their responses ranged between 88% and 97% correct." Can your classifier outperform the human subjects?



## Step 1: Load the Data

The notebook comes pre-bundled with the [Mines vs Rocks data set](https://www.kaggle.com/datasets/mattcarter865/mines-vs-rocks). Our first step is to create a pandas DataFrame from the CSV file. Note that the CSV file has no header row. Loading the CSV file into a DataFrame will make it easy for us to explore the data, preprocess it, and split it into training and test sets.


In [43]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

sonar_csv_path = "../input/mines-vs-rocks/sonar.all-data.csv"
sonar_data = pd.read_csv(sonar_csv_path, header=None)

We now have a pandas DataFrame encapsulating the sonar data, and can proceed with our data exploration.

## Step 2: Explore the Data

**to explore the data we will use the head, description min and max, and shape functions as seen below:**


Lets start out by looking at the first few lines of data

In [44]:
sonar_data.head(n=10)

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
5  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   
6  0.0317  0.0956  0.1321  0.1408  0.1674  0.1710  0.0731  0.1401  0.2083   
7  0.0519  0.0548  0.0842  0.0319  0.1158  0.0922  0.1027  0.0613  0.1465   
8  0.0223  0.0375  0.0484  0.0475  0.0647  0.0591  0.0753  0.0098  0.0684   
9  0.0164  0.0173  0.0347  0.0070  0.0187  0.0671  0.1056  0.0697  0.0962   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
5  0.3039  ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027   
6  0.3513  ...  0.0201  0.0248  0.0131  0.0070  0.0138  0.0092  0.0143   
7  0.2838  ...  0.0081  0.0120  0.0045  0.0121  0.0097  0.0085  0.0047   
8  0.1487  ...  0.0145  0.0128  0.0145  0.0058  0.0049  0.0065  0.0093   
9  0.0251  ...  0.0090  0.0223  0.0179  0.0084  0.0068  0.0032  0.0035   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  
5  0.0051  0.0062   R  
6  0.0036  0.0103   R  
7  0.0048  0.0053   R  
8  0.0059  0.0022   R  
9  0.0056  0.0040   R  

[10 rows x 61 columns]

We can explore the data in a similar maner with the describe function and get more summary statistics

In [45]:
sonar_data.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

Lets see the number of rows and colums

In [46]:
sonar_data.shape


(208, 61)

Lets see what the range of data is

In [47]:
print(f"The min value in sonar_data is {min(sonar_data.describe().loc['min'])}\nThe max value in sonar_data is:{max(sonar_data.describe().loc['max'])}")

The min value in sonar_data is 0.0
The max value in sonar_data is:1.0


**We can see this dataset has 61 columns, 59 are numerical dependent variables and 1 independent variable “R” or “M” rock or mine. There are 208 lines of data. Looks like the features of the dataset are already normalized between 0 and 1.**

## Step 3: Preprocess the Data

**This dataset is in good shape since it's already normalized. We should convert the string "R" and "M" to numeric values to aid in prediction. We can use 1 for rock (R) and -1 for mine (M).**

Lets to a look at the dataset before the change

In [48]:
sonar_data.iloc[:,60]

0      R
1      R
2      R
3      R
4      R
      ..
203    M
204    M
205    M
206    M
207    M
Name: 60, Length: 208, dtype: object

we can see the column 60 is string values R and M

Now we can use the numpy where function to conver these values to numerical representations and then we will use the same code as above to confirm the string to int conversion

In [49]:
sonar_data.iloc[:, 60] = np.where(sonar_data.iloc[:, 60] == "R", 1, np.where(sonar_data.iloc[:, 60] == "M", -1, sonar_data.iloc[:, 60]))
sonar_data.iloc[:,60]

0       1
1       1
2       1
3       1
4       1
       ..
203    -1
204    -1
205    -1
206    -1
207    -1
Name: 60, Length: 208, dtype: object

**The results of converting R&M to 1 or - 1 will be useful in order to train the perceptron and adjust the weight biases. We will need to be aware of this change and correct for it back into the R & M format if reporting any results to make them meaningful outcomes of the model.**

## Step 4: Prepare the Training and Test Data Sets
**We can use the train test split function from SK learn to separate the sonar data into a training and testing data set we'll want to be conscious that the output (target) variable is the last variable of the data set and the features are all other columns. The train split function will split that 208 rows of data appropriately and we can use a 80:20 ratio.**

In [50]:
from sklearn.model_selection import train_test_split


# Separate features and target columns
X = sonar_data.iloc[:, :-1]
y = sonar_data.iloc[:, -1]
y = y.astype(int)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

lets do a sanity check and use head and shape to examine our split for size and randomness before proceeding

In [51]:
X_train.head(n=10)

0       1       2       3       4       5       6       7       8   \
86   0.0188  0.0370  0.0953  0.0824  0.0249  0.0488  0.1424  0.1972  0.1873   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
82   0.0409  0.0421  0.0573  0.0130  0.0183  0.1019  0.1054  0.1070  0.2302   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
194  0.0392  0.0108  0.0267  0.0257  0.0410  0.0491  0.1053  0.1690  0.2105   
38   0.0123  0.0022  0.0196  0.0206  0.0180  0.0492  0.0033  0.0398  0.0791   
24   0.0293  0.0644  0.0390  0.0173  0.0476  0.0816  0.0993  0.0315  0.0736   
60   0.0130  0.0006  0.0088  0.0456  0.0525  0.0778  0.0931  0.0941  0.1711   
195  0.0129  0.0141  0.0309  0.0375  0.0767  0.0787  0.0662  0.1108  0.1777   

         9   ...      50      51      52      53      54      55      56  \
86   0.1806  ...  0.0143  0.0093  0.0033  0.0113  0.0030  0.0057  0.0090   
203  0.2684  ...  0.0203  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065   
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
82   0.2259  ...  0.0113  0.0028  0.0036  0.0105  0.0120  0.0087  0.0061   
205  0.2529  ...  0.0155  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140   
194  0.2471  ...  0.0089  0.0083  0.0080  0.0026  0.0079  0.0042  0.0071   
38   0.0475  ...  0.0149  0.0125  0.0134  0.0026  0.0038  0.0018  0.0113   
24   0.0860  ...  0.0170  0.0035  0.0052  0.0083  0.0078  0.0075  0.0105   
60   0.1483  ...  0.0092  0.0078  0.0041  0.0013  0.0011  0.0045  0.0039   
195  0.2245  ...  0.0204  0.0124  0.0093  0.0072  0.0019  0.0027  0.0054   

         57      58      59  
86   0.0057  0.0068  0.0024  
203  0.0115  0.0193  0.0157  
67   0.0067  0.0078  0.0068  
82   0.0061  0.0030  0.0078  
205  0.0138  0.0077  0.0031  
194  0.0044  0.0022  0.0014  
38   0.0058  0.0047  0.0071  
24   0.0160  0.0095  0.0011  
60   0.0022  0.0023  0.0016  
195  0.0017  0.0024  0.0029  

[10 rows x 60 columns]

In [52]:
X_train.shape

(166, 60)

In [53]:
X_test.head(n=10)

0       1       2       3       4       5       6       7       8   \
161  0.0305  0.0363  0.0214  0.0227  0.0456  0.0665  0.0939  0.0972  0.2535   
15   0.0298  0.0615  0.0650  0.0921  0.1615  0.2294  0.2176  0.2033  0.1459   
73   0.0139  0.0222  0.0089  0.0108  0.0215  0.0136  0.0659  0.0954  0.0786   
96   0.0181  0.0146  0.0026  0.0141  0.0421  0.0473  0.0361  0.0741  0.1398   
166  0.0411  0.0277  0.0604  0.0525  0.0489  0.0385  0.0611  0.1117  0.1237   
9    0.0164  0.0173  0.0347  0.0070  0.0187  0.0671  0.1056  0.0697  0.0962   
100  0.0629  0.1065  0.1526  0.1229  0.1437  0.1190  0.0884  0.0907  0.2107   
135  0.0094  0.0611  0.1136  0.1203  0.0403  0.1227  0.2495  0.4566  0.6587   
18   0.0270  0.0092  0.0145  0.0278  0.0412  0.0757  0.1026  0.1138  0.0794   
148  0.0712  0.0901  0.1276  0.1497  0.1284  0.1165  0.1285  0.1684  0.1830   

         9   ...      50      51      52      53      54      55      56  \
161  0.3127  ...  0.0271  0.0200  0.0070  0.0070  0.0086  0.0089  0.0074   
15   0.0852  ...  0.0154  0.0031  0.0153  0.0071  0.0212  0.0076  0.0152   
73   0.1015  ...  0.0024  0.0062  0.0072  0.0113  0.0012  0.0022  0.0025   
96   0.1045  ...  0.0076  0.0223  0.0255  0.0145  0.0233  0.0041  0.0018   
166  0.2300  ...  0.0181  0.0217  0.0038  0.0019  0.0065  0.0132  0.0108   
9    0.0251  ...  0.0118  0.0090  0.0223  0.0179  0.0084  0.0068  0.0032   
100  0.3597  ...  0.0257  0.0089  0.0262  0.0108  0.0138  0.0187  0.0230   
135  0.5079  ...  0.0480  0.0234  0.0175  0.0352  0.0158  0.0326  0.0201   
18   0.1520  ...  0.0045  0.0084  0.0010  0.0018  0.0068  0.0039  0.0120   
148  0.2127  ...  0.0154  0.0154  0.0156  0.0054  0.0030  0.0048  0.0087   

         57      58      59  
161  0.0042  0.0055  0.0021  
15   0.0049  0.0200  0.0073  
73   0.0059  0.0039  0.0048  
96   0.0048  0.0089  0.0085  
166  0.0050  0.0085  0.0044  
9    0.0035  0.0056  0.0040  
100  0.0057  0.0113  0.0131  
135  0.0168  0.0245  0.0154  
18   0.0132  0.0070  0.0088  
148  0.0101  0.0095  0.0068  

[10 rows x 60 columns]

In [54]:
X_test.shape

(42, 60)

In [55]:
y_test[0:9]

161   -1
15     1
73     1
96     1
166   -1
9      1
100   -1
135   -1
18     1
Name: 60, dtype: int64

In [56]:
y_test.shape

(42,)

In [57]:
y_train[0:9]

86     1
203   -1
67     1
82     1
205   -1
194   -1
38     1
24     1
60     1
Name: 60, dtype: int64

**We successfully split the training and test data based on 80:20 rule and divided the feature verse outcome.  We verified it this split and saw that X_train has 166 values and X_test has 42 rows the same distribution occurred with the Y test and train. The X data features the 59 records and the y data is only 1 and - 1 outcomes as we expect.**

## Step 5: Instantiate and Configure a Perceptron
**In this step we will initiate and configure a perceptron model from SK learn and define a few of the parameters within the model**

In [58]:
from sklearn.linear_model import Perceptron
#help(Perceptron)
perceptron = Perceptron(alpha = 0.1, max_iter = 10000)

**following prior examples we set alpha to .01 in the max_itter or number of epochs to 10,000 this should serve as good base model parameters but if convergence does not occur we can expand the number of epochs or tweak the settings otherwise of the model**

## Step 6: Train the Perceptron
**We will train the perceptron model with the X and y train datasets using the fit metho**

In [59]:
# Convert varriables to inttype
y_train = y_train.astype(int)
y_test = y_test.astype(int)

In [60]:
#help(perceptron.fit)
perceptron.fit(X_train, y_train)

Perceptron(alpha=0.1, max_iter=10000)

**we discovered that the variable types needed to be converted to integers prior to training the model otherwise the perceptron fit method worked just fine given our parameters of alpha and number of epochs**

## Step 7: Test the Perceptron

**In this section we will test the perceptron using the score function and report the model's accuracy on the test data set**


In [61]:
accuracy = perceptron.score(X_test, y_test)
print(f'Test accuracy: {accuracy * 100:.1f}%')

Test accuracy: 88.1%


**We can see that the accuracy of the perceptron is 88.1% on the test data set this seems like a  reasonable number for a perceptron model as high 90’s would be a read flag, and of course very low numbers would indicate problems**

## Step 8: Demonstrate Making Predictions
**Here we can use the predict function to convert the numeric features of a small number of X_test data into outcomes and ultimately return what they're not it is classified as Rock or Mine**

In [62]:
results = perceptron.predict(X_test.head(n=10))
for i in results:
    rock_mine_pred = np.where(results == 1, "Rock", np.where(results == -1, "Mine", results))
print(rock_mine_pred)

['Mine' 'Rock' 'Rock' 'Rock' 'Mine' 'Rock' 'Mine' 'Mine' 'Rock' 'Mine']


**Here we used the first 10 lines in the X_test data set to demonstrate the perceptrons prediction ability based off the features of the test data set after we perform this prediction we wanted to convert the numeric results back to the text Rock or Mine values and print those**

## Step 9: Evaluate (and Improve?)

Describe the configuration and performance of your classifier, and what the results mean. Is this a good classifier? Why isn't it better than what it is? What might you try next to improve it? How accurate can you make your classifier? If you have time, see if you can increase the accuracy. Can you beat 98%? 🙀

**The classifier is configured as a linear model with alpha of 0.01 and maximum number of epochs set to 10,000.  The classifier worked well given the 88.1% accuracy. There is an limitation to the accuracy that can be achieved and trying to improve this through hyper parameterization since a linear classifier was chosen if a different type of classifier was chosen a greater accuracy could potentially be achieved through hyper parameterization improving the accuracy was achieved only by reducing the features of the dataset to 30. Manipulating the training test split ratio changed the accuracy similar to cross validation but greater than 88.1% could not be achieved. No beating 98% is not possible given the model selected.**

Can we Manipulating the alpha paramater to improve the model

In [75]:
from sklearn.linear_model import Perceptron

optimized_perceptron = Perceptron(alpha=0.0001, max_iter=10000, penalty= 'none')
optimized_perceptron.fit(X_train, y_train)
optimize_accuracy = optimized_perceptron.score(X_test, y_test)
print(f'Test accuracy: {optimize_accuracy  * 100:.1f}%') #Test accuracy: 88.1%

optimized_perceptron = Perceptron(alpha=0.9, max_iter=10000, penalty= 'none')
optimized_perceptron.fit(X_train, y_train)
optimize_accuracy = optimized_perceptron.score(X_test, y_test)
print(f'Test accuracy: {optimize_accuracy  * 100:.1f}%') #Test accuracy: 88.1% 

Test accuracy: 88.1%
Test accuracy: 88.1%


We can see the range of 0.0001 to 0.9 for alpha did not change the models accuracy

Can we Manipulating the max_iter paramater to improve the model

In [76]:
from sklearn.linear_model import Perceptron
optimized_perceptron = Perceptron(alpha=0.1, max_iter=1000, penalty= 'none')
optimized_perceptron.fit(X_train, y_train)
optimize_accuracy = optimized_perceptron.score(X_test, y_test)
print(f'Test accuracy: {optimize_accuracy  * 100:.1f}%') #Test accuracy: 88.1%

optimized_perceptron = Perceptron(alpha=0.1, max_iter=1000000, penalty= 'none')
optimized_perceptron.fit(X_train, y_train)
optimize_accuracy = optimized_perceptron.score(X_test, y_test)
print(f'Test accuracy: {optimize_accuracy  * 100:.1f}%') #Test accuracy: 88.1%

Test accuracy: 88.1%
Test accuracy: 88.1%


We can see the range of 1000 to 1000000 for max_iter did not change the models accuracy

Can reducing the number of features improve accuracy?

In [67]:
from sklearn.feature_selection import RFE, SelectKBest, chi2
from sklearn.model_selection import train_test_split

# Using SelectKBest
kbest_selector = SelectKBest(chi2, k=30)
X_kbest = kbest_selector.fit_transform(X, y)


# Split the data into training and testing sets
OP_X_train, OP_X_test, OP_y_train, OP_y_test = train_test_split(X_kbest, y, test_size=0.2, random_state=42)

In [68]:
from sklearn.linear_model import Perceptron
#help(Perceptron)
optimized_perceptron = Perceptron(alpha=0.01, max_iter=10000, penalty= 'none')
#help(perceptron.fit)
optimized_perceptron.fit(OP_X_train, OP_y_train)
optimize_accuracy = optimized_perceptron.score(OP_X_test, OP_y_test)
print(f'Test accuracy: {optimize_accuracy  * 100:.1f}%')

Test accuracy: 83.3%


Reducing the number of features to the 30 most imprortant improved the accuracy by only a couple precent.

Could changing the trainging and test split using cross validation work to improve accuracy?

In [69]:
from sklearn.model_selection import cross_val_score

# Instantiate the perceptron model
perceptron = Perceptron(alpha=0.1, max_iter=100000)

# Perform 5-fold cross-validation
cv_scores = cross_val_score(perceptron, X, y, cv=10)

# Print cross-validation scores and average
print(f"Cross-validation scores: {cv_scores}")
print(f"Average cross-validation score: {cv_scores.mean() * 100:.1f}%")

Cross-validation scores: [0.23809524 0.76190476 0.42857143 0.47619048 0.76190476 0.76190476
 0.61904762 0.85714286 0.55       0.65      ]
Average cross-validation score: 61.0%


No this decreased accuracy.  

Give us a quick recap of what you've done here in **Part 2**. Mention _three_ things that were most notable in this process, whether it's related to exploration, preprocessing, configuring, training, or evaluating. If you put in the effort to try to improve the perceptron, describe what you did and what let you to try what you did, and describe the results. Conclude with some questions about the problem or what you might do next to increase the performance of your classifier.

**In Part 2 we took a data set with 59 features and one predictor outcome rock or mine, after exploring the data set in minor preprocessing we split the data into test and training sets and then configured, trained, and tested our perceptron resulting in 88% accuracy and showed its ability to make classifications based off  features from the dataset. Find Tuneing the methods parameters did not improve accuracy given the linear model selected.**

**The first notable thing to me is how you can implement a seemingly complex model with so few lines of code given the availability of packages. That the functions in these packages can complete what would seemingly be a daunting programming challenge to a novice and simply a few lines of code when calling them. The other notable thing is how well the linear classifier worked resulting in a 88% accuracy on what looked like a fairly complex data set with 59 features I'd imagine this is a simple ML model and I'm impressed at this level of accuracy. Finally it was notable to me that I couldn't really improve the accuracy of the base configuration which in hindsight makes sense given it's a linear model but was surprised to see that the default settings essentially was trying to fight through hyper parameterization.**


In [77]:
#help(perceptron)